source/common/config/xds_mux/sotw_subscription_state.cc

```diff
@@ -71,14 +73,69 @@ void SotwSubscriptionState::handleGoodResponse(
   // TODO (dmitri-d) to eliminate decoding of resources twice consider expanding the interface to
   // support passing of decoded resources. This would also avoid a resource copy above.
   callbacks().onConfigUpdate(non_heartbeat_resources, message.version_info());
-  // TODO(abeyad): Add support for calling the Config::UpdateListener registered listeners.
   // Now that we're passed onConfigUpdate() without an exception thrown, we know we're good.
   last_good_version_info_ = message.version_info();
   last_good_nonce_ = message.nonce();
+
+  // Send the resources to the xDS delegate, if configured.
+  if (xds_resources_delegate_.has_value()) {
+    XdsConfigSourceId source_id{target_xds_authority_, message.type_url()};
+    std::vector<DecodedResourceRef> resource_refs;
+    resource_refs.reserve(non_heartbeat_resources.size());
+    for (const DecodedResourcePtr& r : non_heartbeat_resources) {
+      resource_refs.emplace_back(*r);
+    }
+    xds_resources_delegate_->onConfigUpdated(source_id, resource_refs);
+  }
+
   ENVOY_LOG(debug, "Config update for {} (version {}) accepted with {} resources", typeUrl(),
             message.version_info(), message.resources().size());
 }
 
+void SotwSubscriptionState::handleEstablishmentFailure() {
+  BaseSubscriptionState::handleEstablishmentFailure();
+
+  if (previously_fetched_data_ || !xds_resources_delegate_.has_value()) {
+    return;
+  }
+
+  const XdsConfigSourceId source_id{target_xds_authority_, type_url_};
+  TRY_ASSERT_MAIN_THREAD {
+    const std::vector<std::string> resource_names{names_tracked_.begin(), names_tracked_.end()};
+    std::vector<envoy::service::discovery::v3::Resource> resources =
+        xds_resources_delegate_->getResources(source_id, resource_names);
+
+    std::vector<DecodedResourcePtr> decoded_resources;
+    const auto scoped_update = ttl_.scopedTtlUpdate();
+    std::string version_info;
+    for (const auto& resource : resources) {
+      if (version_info.empty()) {
+        version_info = resource.version();
+      } else {
+        ASSERT(version_info == resource.version());
+      }
+
+      TRY_ASSERT_MAIN_THREAD {
+        auto decoded_resource = DecodedResourceImpl::fromResource(*resource_decoder_, resource);
+        setResourceTtl(*decoded_resource);
+        decoded_resources.emplace_back(std::move(decoded_resource));
+      }
+      END_TRY
+      catch (const EnvoyException& e) {
+        xds_resources_delegate_->onResourceLoadFailed(source_id, resource.name(), e);
+      }
+    }
+
+    callbacks().onConfigUpdate(decoded_resources, version_info);
+  }
```

Comment:
What should be the values of `last_good_version_info_` and `last_good_nonce_` if this ingestion is successful?

Reply:
I'm interested in your thoughts on that.  I didn't think I should touch `last_good_version_info_` and `last_good_nonce_` as part of loading cached resources, since those variables are used in request/response interactions with the xDS server.  Do you believe that the cached config should set those values (specifically the `last_good_version_info_` value)?  If so, why?

Reply:
I think it impacts what will be sent to the server on the [next request](https://github.com/envoyproxy/envoy/blob/ad186c448cf8c5e78454d2adf5f228e9c90a997c/source/common/config/xds_mux/sotw_subscription_state.cc#L88-L90), so it can avoid the response, if the version hasn't changed.

Reply:
My thought was, even if the version is the same, once the management server can be reached again, we should send a request, even if the version is the latest one.  Would resending the same nonce make sense, even if we sent over the cached version number in the next request?

Reply:
Reading https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#basic-protocol-overview, it seems like a nonce does not survive stream restarts/reconnects, so I assume it's already possible to send a non-empty version info with an empty nonce in a DiscoveryRequest?  If that's the case, I can do the same with using the persisted config.  Is my understanding correct?

Reply:
The nonce is reset, but not the version.
The idea is that by sending the version, the server is able to know if the client already holds the latest and greatest, and if it does, the server won't need to send an update.
Not sure if your use-case fits should also follow this.

Reply:
I think it's tricky.  I'm imagining a couple scenarios: 

1. What if the DiscoveryRequest requests resources {"a", "b", "c"}, but the xDS delegate's getResources() function only had {"a", "b"} cached, but not "c".  Do we still want to set the last known good version?  Or is it guaranteed that, if there's a new resource added to the list (in this case, "c"), the resources would have a new version too?

2. What if the xDS delegate returned {"a", "b", "c"}, but "c" was malformed (perhaps an invalid enum value) and so doesn't get used.  Should we still set the last known good version to the version found in "a" and "b"?

Reply:
It seems like the version always gets updated if any resource gets changed/added/removed for that resource type: https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack-nack-and-resource-type-instance-version.  I pushed https://github.com/envoyproxy/envoy/pull/23281/commits/08722407f5e6adb6283087e167d1b930c8d04b7b which addresses this scenario. PTAL.